### PR TITLE
WFLY-19903 Upgrade wildfly-clustering to 4.0.4.Final

### PR DESCRIPTION
**Pull request by pferraro**
Incorporates:
WFLY-19904 Upgrade JGroups to 5.3.x
WFLY-19905 Upgrade Infinispan to 15.0.x

https://issues.redhat.com/browse/WFLY-19903
https://issues.redhat.com/browse/WFLY-19904
https://issues.redhat.com/browse/WFLY-19905

### CODE REVIEW

**Comment by scottmarlow** [comment]
We are using Hibernate ORM 6.6 in WildFly 35 and looking to integrate Hibernate ORM 7.0 in WildFly Preview (possibly for 35).

Do we know if Infinispan 15.0.x works with both Hibernate ORM 6.x + 7.0?

**Review by rhusar — CHANGES_REQUESTED** [review]
The `org.jboss.as.test.clustering.cluster.AbstractClusteringTestCase#INFINISPAN_SERVER_PROFILE_DEFAULT` needs to be updated.

**Comment by pferraro** [comment]
> The `org.jboss.as.test.clustering.cluster.AbstractClusteringTestCase#INFINISPAN_SERVER_PROFILE_DEFAULT` needs to be updated.

Already done.

**Comment by pferraro** [comment]
> We are using Hibernate ORM 6.6 in WildFly 35 and looking to integrate Hibernate ORM 7.0 in WildFly Preview (possibly for 35).
> 
> Do we know if Infinispan 15.0.x works with both Hibernate ORM 6.x + 7.0?

@scottmarlow I am not aware of any second-level cache SPI changes that would make 7.0 incompatible with the existing implementation.  If any exist, they should become obvious fairly quickly, in which case, we'll request a new implementation module.

**Comment by rhusar** [comment]
Sticking hold until we figure out the test failure

```
[ERROR] Tests run: 12, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 30.786 s <<< FAILURE! - in org.jboss.as.test.clustering.single.infinispan.InfinispanModulesTestSuite
[ERROR] testIndexed(org.jboss.as.test.clustering.single.infinispan.query.RemoteQueryTestCase)  Time elapsed: 0.752 s  <<< ERROR!
org.infinispan.client.hotrod.exceptions.HotRodClientException:Request for messageId=26 returned server error (status=0x85): org.infinispan.commons.CacheException: Could not load codec 'Lucene87'. Did you forget to add lucene-backward-codecs.jar?
java.lang.IllegalArgumentException: Could not load codec 'Lucene87'. Did you forget to add lucene-backward-codecs.jar?
java.lang.IllegalArgumentException: An SPI class of type org.apache.lucene.codecs.Codec with name 'Lucene87' does not exist.  You need to add the corresponding JAR file supporting this SPI to your classpath.  The current classpath supports the following names: [Lucene99]
```

**Comment by bstansberry** [comment]
@rhusar @pferraro @scottmarlow FYI RemoteQueryTestCase.testIndexed failed on #18354 as well, although the failure didn't say anything about Lucene.

**Comment by rhusar** [comment]
> @rhusar @pferraro @scottmarlow FYI RemoteQueryTestCase.testIndexed failed on #18354 as well, although the failure didn't say anything about Lucene.

Thanks for pointing that out, the 'CacheException: indexCreatedVersionMajor is in the future: 9' I believe is a result of the index being written in permanent location on the CI hosts and is not purged. Part of the fix includes overriding Infinispan default location to a temporary build directory.

**Comment by bstansberry** [comment]
@rhusar Great; glad that helped. Just FYI it happened again at https://ci.wildfly.org/viewLog.html?buildId=466016&buildTypeId=WF_PullRequest_GalleonLinuxJdk17&fromSakuraUI=true.

File: ee-feature-pack/galleon-shared/src/main/resources/modules/system/layers/base/io/github/resilience4j/core/main/module.xml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright The WildFly Authors
+  ~ SPDX-License-Identifier: Apache-2.0
+  -->
+<module xmlns="urn:jboss:module:1.9" name="io.github.resilience4j.core">
+
```
Comment (bstansberry) [inline ee-feature-pack/galleon-shared/src/main/resources/modules/system/layers/base/io/github/resilience4j/core/main/module.xml:7]: jboss-api = private?

File: ee-feature-pack/galleon-shared/src/main/resources/modules/system/layers/base/io/github/resilience4j/retry/main/module.xml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright The WildFly Authors
+  ~ SPDX-License-Identifier: Apache-2.0
+  -->
+<module xmlns="urn:jboss:module:1.9" name="io.github.resilience4j.retry">
+
```
Comment (bstansberry) [inline ee-feature-pack/galleon-shared/src/main/resources/modules/system/layers/base/io/github/resilience4j/retry/main/module.xml:7]: jboss-api = private?

File: ee-feature-pack/galleon-shared/src/main/resources/modules/system/layers/base/io/github/resilience4j/core/main/module.xml (hunk 1 above)
Comment (pferraro) [reply ee-feature-pack/galleon-shared/src/main/resources/modules/system/layers/base/io/github/resilience4j/core/main/module.xml:7]: Whoops. Yes, this module should be private.

Comment (pferraro) [reply ee-feature-pack/galleon-shared/src/main/resources/modules/system/layers/base/io/github/resilience4j/core/main/module.xml:7]: Fixed.

File: ee-feature-pack/galleon-shared/src/main/resources/modules/system/layers/base/io/github/resilience4j/retry/main/module.xml (hunk 2 above)
Comment (pferraro) [reply ee-feature-pack/galleon-shared/src/main/resources/modules/system/layers/base/io/github/resilience4j/retry/main/module.xml:7]: Fixed.

**Review by rhusar — CHANGES_REQUESTED** [review]
There are still outstanding security manager issues that need to be resolved.

**Review by rhusar — CHANGES_REQUESTED** [review]
There are still outstanding security manager issues (I cannot comment on actual lines so leaving a generic comment).

**Review by bstansberry — APPROVED** [review]

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/marshalling/UserMarshallerFactory.java [hunk 3]
```diff
@@ -41,7 +41,26 @@
...
+            }
+
+            @Override
+            public String getProtoFile() throws UncheckedIOException {
+                return null;
+            }
+
+            @Override
+            public void registerSchema(SerializationContext context) {
+                org.infinispan.query.remote.client.impl.MarshallerRegistration.init(context);
```
Comment (rhusar) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/marshalling/UserMarshallerFactory.java:57]: Nitpick: no reason to use FQCN here.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/persistence/hotrod/HotRodStoreConfiguration.java [hunk 4]
```diff
@@ -18,14 +18,14 @@
...
-public class HotRodStoreConfiguration extends AbstractStoreConfiguration {
+public class HotRodStoreConfiguration extends AbstractStoreConfiguration<HotRodStoreConfiguration> {
 
     static final AttributeDefinition<RemoteCacheContainer> REMOTE_CACHE_CONTAINER = AttributeDefinition.builder("remoteCacheContainer", null, RemoteCacheContainer.class).build();
 
     static final AttributeDefinition<String> CACHE_CONFIGURATION = AttributeDefinition.builder("cacheConfiguration", null, String.class).build();
 
     public HotRodStoreConfiguration(AttributeSet attributes, AsyncStoreConfiguration async) {
-        super(attributes, async);
+        super(org.infinispan.persistence.remote.configuration.Element.REMOTE_STORE, attributes, async);
```
Comment (rhusar) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/persistence/hotrod/HotRodStoreConfiguration.java:28]: Nitpick: no reason to use FQCN here.

**Review by rhusar — APPROVED** [review]
Overall LGTM!

**Comment by bstansberry** [comment]
The failed 'Linux - JDK 21' job was due to an environmental problem (downloading infinispan-server) and had passed against the same PR sha at https://ci.wildfly.org/viewLog.html?buildId=471674. So I'm good for merging this.

**Comment by bstansberry** [comment]
Thanks @pferraro and @rhusar

**Comment by rhusar** [comment]
> The failed 'Linux - JDK 21' job was due to an environmental problem (downloading infinispan-server) and had passed against the same PR sha at https://ci.wildfly.org/viewLog.html?buildId=471674. So I'm good for merging this.

Yes, that does happen from time to time when the download server is unavailable and not cached (which easily happens on an upgrade like this).

**Pull request merged**